### DISCUSSION TITLE
Need way to securely verify Internet Archive crawler

### DISCUSSION THREAD

**Discussion by mlissner** [internetarchive/heritrix3#507] [Q&A]
I could be mistaken, but I haven't been able to find the documentation on how to verify the IA crawler. The closest thing I've found is that you can check the User-Agent string of the crawler, but that's easily faked. My issue is that I want to invite the IA crawler to crawl my content, but I want to detect things like spammers and block them.

Google and Bing both handle this by using a reverse DNS request of the IP address of the crawler, followed by a regular DNS request checking the host returned by the reverse DNS. 

Put another way:

    host 66.249.66.1
    1.66.249.66.in-addr.arpa domain name pointer crawl-66-249-66-1.googlebot.com.

    host crawl-66-249-66-1.googlebot.com
    crawl-66-249-66-1.googlebot.com has address 66.249.66.1

So, since the second host command returns the same IP that we started with, and since the domain ends with googlebot.com, we're in business.

Here's google's docs: https://support.google.com/webmasters/answer/80553?hl=en
And Bings: https://www.bing.com/webmaster/help/how-to-verify-bingbot-3905dc26

Could IA add this feature too? I think it would only require that you do some work with your DNS whenever you have a new IP address.

**Comment by nlevitt** [comment]
Sorry, seeing this for the first time.

I think your DNS technique may already work for our crawlers?

**Reply by ansuz** [reply]
I've received requests from the following IP addresses:

* 152.53.36.192
* 152.53.39.118
* 2607:5300:60:6dd4::
* 2a03:4000:5c:152::
* 94.16.31.119

identified by this user-agent string:

`ArchiveTeam ArchiveBot/20250806.050c783 (wpull 2.0.3) and not Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.90 Safari/537.36`

A reverse DNS lookup yields the following results:

```
$ for ip in 152.53.36.192 152.53.39.118 2607:5300:60:6dd4:: 2a03:4000:5c:152:: 94.16.31.119;do echo $ip;host $ip;echo;done;
152.53.36.192
192.36.53.152.in-addr.arpa domain name pointer v2202411233004298790.luckysrv.de.

152.53.39.118
118.39.53.152.in-addr.arpa domain name pointer v2202411243627299646.bestsrv.de.

2607:5300:60:6dd4::
Host 0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.4.d.d.6.0.6.0.0.0.0.3.5.7.0.6.2.ip6.arpa not found: 3(NXDOMAIN)

2a03:4000:5c:152::
Host 0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.2.5.1.0.c.5.0.0.0.0.0.4.3.0.a.2.ip6.arpa not found: 3(NXDOMAIN)

94.16.31.119
Host 119.31.16.94.in-addr.arpa. not found: 3(NXDOMAIN)
```

Can you confirm whether this is evidence of others spoofing their agent string to pass as ArchiveTeam bots?

If so, could you point to an IP address of a valid ArchiveBot crawler so I could validate the method against a legitimate example?

**Reply by ansuz** [reply]
**Update**: I found the following user-agent strings in my logs:

```
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/0458982 warc/v0.8.90"
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/22cbc49 warc/v0.8.90"
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/2abbcd3 warc/v0.8.90"
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/a07610d warc/v0.8.85"
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/a9fc8de warc/v0.8.78"
"Mozilla/5.0 (compatible; archive.org_bot +http://archive.org/details/archive.org_bot) Zeno/f160c5b warc/v0.8.85"
```

...and confirmed that the rDNS validation method works as expected, with IPs resolving to records of the form `crawl911.us.archive.org.`.

A smaller number of log entries instead set their user-agent string to something like:

```
"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36"
```

...and instead identify themselves using a `via` header:

```
"Mozilla/5.0 (compatible; archive.org_bot; Wayback Machine Live Record; http://archive.org/details/archive.org_bot), Mozilla/5.0 (compatible; archive.org_bot; Wayback Machine Live Record; http://archive.org/details/archive.org_bot), 1.1 warcprox"
```

The rDNS records for these source IPs still validate as expected, though due to this somewhat uncommon usage I've had to make an exception for how such archive.org bots are handled in order to automate bot validation.

Also, I'm still not sure whether those ArchiveTeam bots mentioned above are being spoofed.

**Comment by mlissner** [comment]
Would it be possible to get this documented somewhere? I'd be afraid to start using this technique without some level of commitment from IA, for fear that it'd break your crawler.

**Comment by TheDocTrier** [comment]
I am unsure of the relationship between this project and the Wayback machine, but I just tested the Wayback on my own site and recorded the resulting access log:

```sh
207.241.225.58 - - [03/Mar/2026:08:07:56 +0000] "GET /does_not_exist/ HTTP/1.1" 404 1564 "-" "python-urllib3/2.2.3" "-"
207.241.238.29 - - [03/Mar/2026:08:07:58 +0000] "GET /does_not_exist/ HTTP/1.1" 404 4008 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
207.241.225.90 - - [03/Mar/2026:08:07:58 +0000] "GET /style.css HTTP/1.1" 200 32630 "https://brooksb.dev/does_not_exist/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
2620:0:9c0:4:: - - [03/Mar/2026:08:07:58 +0000] "GET /basic/favicon.svg HTTP/1.1" 200 310 "https://brooksb.dev/does_not_exist/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
207.241.234.144 - - [03/Mar/2026:08:07:59 +0000] "GET /basic/by-sa.svg HTTP/1.1" 200 6181 "https://brooksb.dev/does_not_exist/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
207.241.238.29 - - [03/Mar/2026:08:08:00 +0000] "GET /basic/favicon.png HTTP/1.1" 200 3139 "https://brooksb.dev/does_not_exist/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
2620:0:9c0:20:: - - [03/Mar/2026:08:08:00 +0000] "GET /basic/favicon.svg HTTP/1.1" 200 310 "https://brooksb.dev/does_not_exist/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36" "-"
```

The technique mentioned above worked for the associated IPv4 addresses:

```text
58.225.241.207.in-addr.arpa domain name pointer wwwb-spn35.us.archive.org.
29.238.241.207.in-addr.arpa domain name pointer wwwb-cdxredis0.us.archive.org.
90.225.241.207.in-addr.arpa domain name pointer wwwb-spn33.us.archive.org.
144.234.241.207.in-addr.arpa domain name pointer wwwb-front8.us.archive.org.
29.238.241.207.in-addr.arpa domain name pointer wwwb-cdxredis0.us.archive.org.

wwwb-spn35.us.archive.org has address 207.241.225.58
wwwb-cdxredis0.us.archive.org has address 207.241.238.29
wwwb-spn33.us.archive.org has address 207.241.225.90
wwwb-front8.us.archive.org has address 207.241.234.144
wwwb-cdxredis0.us.archive.org has address 207.241.238.29
```

However, similarly to @ansuz , it failed for the IPv6 addresses:

```text
Host 0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.4.0.0.0.0.c.9.0.0.0.0.0.0.2.6.2.ip6.arpa not found: 2(SERVFAIL)
Host 0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.2.0.0.0.c.9.0.0.0.0.0.0.2.6.2.ip6.arpa not found: 2(SERVFAIL)
```

As mentioned by @mlissner , it would be helpful to have this implemented (including for IPv6) and documented somewhere so that we can verify crawler requests and be confident that this functionality will not break in the future. This may also be particularly important since the crawler does not always identify itself in the user agent.
